qiita_db/download_link.py

```diff
@@ -0,0 +1,130 @@
+# -----------------------------------------------------------------------------
+# Copyright (c) 2014--, The Qiita Development Team.
+#
+# Distributed under the terms of the BSD 3-clause License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# -----------------------------------------------------------------------------
+
+from __future__ import division
+
+import qiita_db as qdb
+
+from jose import jwt as jose_jwt
+from datetime import datetime, timezone
+from qiita_core.qiita_settings import qiita_config
+
+
+class DownloadLink(qdb.base.QiitaObject):
+    r"""
+    A shortened url for downloading artifacts
+    alongside a signed jwt and expiration
+
+    Methods
+    -------
+    delete_expired
+
+    See Also
+    --------
+    qiita_db.QiitaObject
+    """
+
+    _table = "download_link"
+
+    @classmethod
+    def create(cls, jwt):
+        r"""Creates a new object with a new id on the storage system
+
+        Parameters
+        ----------
+        jwt : Json Web Token signing the access link.
+        This jwt will have, at a minimum, jti and exp fields
+
+        Raises
+        ------
+        IncompetentQiitaDeveloperError
+            If the jwt is improperly signed or doesn't contain a jti or exp
+        QiitaDBDuplicateError
+            If the jti already exists in the database
+        """
+
+        jwt_data = jose_jwt.decode(jwt,
+                                   qiita_config.jwt_secret,
+                                   algorithms='HS256')
+        jti = jwt_data["jti"]
+        exp = datetime.utcfromtimestamp(jwt_data["exp"] / 1000)
+
+        with qdb.sql_connection.TRN:
+            if cls.exists(jti):
+                raise qdb.exceptions.QiitaDBDuplicateError(
+                    "JTI Already Exists")
+
+            # Insert study into database
+            sql = """INSERT INTO qiita.{0} (jti, jwt, exp)
+            VALUES (%s, %s, %s) RETURNING jti""".format(cls._table)
+            qdb.sql_connection.TRN.add(sql, [jti, jwt, exp])
+            return qdb.sql_connection.TRN.execute_fetchlast()
+
+    @classmethod
+    def delete(cls, jti):
+        r"""Deletes the link with specified jti from the storage system
+
+        Parameters
+        ----------
+        jti : object
+            The jwt token identifier
+
+        """
+        with qdb.sql_connection.TRN:
+            sql = """DELETE FROM qiita.{0} WHERE jti=%s""".format(cls._table)
+            qdb.sql_connection.TRN.add(sql, [jti])
+            return qdb.sql_connection.TRN.execute_fetchlast()
```

Comment:
What's the return value here?

Reply:
I admittedly have no idea how sql_connection.TRN is supposed to be used.  In an ideal world, I would return whether the cursor's row count was 1, indicating that a row was deleted.  I think I'll settle for returning nothing and switch it to a TRN.execute()